[Draft] Solution submission

josojo opened the pull request:
PR only for early revision...
Also, I will split this PR into smaller chunks tomorrow.

This Pr provides the first `solutionSubmission` functionality.

It does the following for sellOrders only:
- updates all touched orders
- updates all balances
- restores old states from previous solutions
- stores updates for recreation of old solutions

it does not calculate any utility, does not check prices conformities or norms

josojo committed:
first draft for solutionSubmission

josojo committed:
all positive tests for settlement for sellorders is written

josojo committed:
first solution with weird sellTokenPriceIndex

josojo committed:
solution with findPriceIndex()

josojo commented on contracts/StablecoinConverter.sol:
```diff
@@ -116,4 +119,147 @@ contract StablecoinConverter is EpochTokenLocker {
+            require(order.validFrom <= batchIndex, "Order is not yet valid");
+            require(order.validUntil >= batchIndex, "Order is no longer valid");
+            require(
+                prices[order.sellToken].mul(order.buyAmount) >= prices[order.buyToken].mul(order.sellAmount),
```
check limit prices with buyVolume and sellVolume

fleupold commented on contracts/StablecoinConverter.sol:
```diff
@@ -116,4 +119,147 @@ contract StablecoinConverter is EpochTokenLocker {
+        uint len = owner.length;
+        for(uint i = 0; i < len; i++){
+            Order memory order = orders[owner[i]][orderId[i]];
+            require(
```
How could this ever be false?

fleupold commented on contracts/StablecoinConverter.sol:
```diff
@@ -116,4 +119,147 @@ contract StablecoinConverter is EpochTokenLocker {
+            uint buyVolume = volume[i].mul(prices[findPriceIndex(order.buyToken, tokenIdForPrice)]) /
+                prices[findPriceIndex(order.sellToken, tokenIdForPrice)];
+            require(order.sellAmount >= sellVolume, "sellVolume bigger than specified in order");
+            orders[owner[i]][orderId[i]].buyAmount = uint128(
```
Not sure where this mix of integer width comes from but should the volumes than also be provided in u128?

fleupold reviewed (commented):
Looking forward to the break-up. This is a lot to digest/review.

josojo committed:
StablecoinConverter with Orders and logic for updateStateIndex (#160)

This Pr implements orders, which are fully stored on-chain. Its implementation is according to the discussion here:
#158

It also adds the functionality of setting the current state index to the EpochTokenLocker. The currentStateIndex will just always be set currentStateIndex = now/300

josojo committed:
merging with current master

josojo commented:
no longer relevant
